File: Week2/과제/완전_탐색/수식_최대화/Solution_Maximum.py
# 1. 시스템 입력(sys) 및 연산자들의 모든 우선순위 순열(permutations)을 생성할 파이썬 표준 라이브러리를 가져옵니다.
import sys
from itertools import permutations


/**
 * [두 숫자와 연산자를 전달받아 사칙연산 결과를 계산해 주는 보조 함수]
 * @param n1 : 계산에 쓰일 첫 번째 숫자
 * @param n2 : 계산에 쓰일 두 번째 숫자
 * @param op : 사용할 사칙연산 기호 문자 ('+', '-', '*' 중 하나)
 * @return   : 연산된 최종 값
 */
def calculate(n1, n2, op):
    # 더하기일 때
    if op == '+':
        return n1 + n2
    # 빼기일 때
    if op == '-':
        return n1 - n2
    # 곱하기일 때
    if op == '*':
        return n1 * n2


/**
 * [수식을 쪼개고 연산자들의 우선순위 조합을 모두 시뮬레이션하여 가장 큰 절대값을 찾는 메인 핵심 함수]
 * @param expression : 원본 수식 문자열 (예: "100-200*300-500+20")
 * @return           : 조합 가능한 우선순위 하에 도출할 수 있는 가장 큰 절대값
 */
def solution(expression):
    nums = []  # 수식에서 추출해낸 정수형 숫자들을 순서대로 담아둘 리스트입니다.
    ops = []   # 수식에서 추출해낸 연산자 기호들을 순서대로 담아둘 리스트입니다.
    temp = ""  # 숫자의 자릿수들을 임시로 덧붙여 보관할 문자열 변수입니다.
    
    # 원본 수식의 각 글자를 한 문자씩 차례대로 훑어봅니다.
    for char in expression:
        # 글자가 숫자라면 (0~9 사이)
        if char.isdigit():
            temp += char  # 임시 자릿수 문자열에 더해줍니다.
        # 글자가 연산자라면
        else:
            nums.append(int(temp)) # 완성된 숫자를 정수(int)로 변환해 숫자 리스트에 집어넣습니다.
            ops.append(char)       # 연산자를 연산자 리스트에 집어넣습니다.
            temp = ""              # 다음 숫자를 담기 위해 임시 보관함을 비웁니다.
            
    # 반복문이 끝나고 마지막에 남아있던 숫자를 마저 리스트에 넣어줍니다.
    nums.append(int(temp))

    max_reward = 0  # 가장 큰 절댓값 결과를 저장할 변수이며, 0으로 시작합니다.
    
    # 2. permutations(['+', '-', '*'])를 사용하여 3가지 연산자의 모든 우선순위 배치(순열)를 조사합니다.
    #    (예: ('+', '-', '*'), ('+', '*', '-'), ... 총 6가지 조합)
    for priority in permutations(['+', '-', '*']):
        # 원본 데이터를 훼손하지 않기 위해 슬라이싱 복사([:])를 이용해 현재 순위 연산용 임시 리스트들을 만듭니다.
        curr_nums = nums[:]
        curr_ops = ops[:]

        # 현재 우선순위 순서 리스트(priority)에 맞게 연산자 하나씩 순서대로 연산을 처리합니다.
        for p_op in priority:
            i = 0
            # 임시 연산자 리스트를 처음부터 끝까지 스캔합니다.
            while i < len(curr_ops):
                # 지금 연산해야 하는 기호(p_op)와 리스트 상의 기호가 맞아떨어질 경우
                if curr_ops[i] == p_op:
                    # i번째 숫자와 바로 다음(i+1)번째 숫자를 가져와 연산 처리합니다.
                    res = calculate(curr_nums[i], curr_nums[i + 1], p_op)
                    
                    # 계산 결과 값을 원래 첫 번째 숫자가 있던 자리에 덮어씁니다.
                    curr_nums[i] = res
                    # 이미 사용한 두 번째 숫자를 리스트에서 날려버립니다.
                    curr_nums.pop(i + 1)
                    # 계산에 쓴 연산자도 리스트에서 삭제합니다.
                    curr_ops.pop(i)
                    # 요소를 삭제했기 때문에 전체 길이가 줄었습니다. 
                    # 따라서 인덱스(i)를 그대로 유지해 다음 연산 대상을 바로 이어서 봅니다.
                else:
                    i += 1  # 연산자가 일치하지 않는다면 다음 자리를 보러 가기 위해 인덱스를 증가시킵니다.
                    
        # 수식 계산을 완전히 마치고 유일하게 남은 결과값(curr_nums[0])의 절대값(abs)을 구합니다.
        # 기존 최댓값과 비교하여 더 높은 값을 선택하고 기록을 갱신합니다.
        max_reward = max(max_reward, abs(curr_nums[0]))
        
    # 모든 연산자 배치 조합을 전부 검사하고 난 후 찾은 최대 절댓값 결과를 반환합니다.
    return max_reward


# 스크립트가 메인 실행일 때 작동하는 진입점입니다.
if __name__ == "__main__":
    try:
        # 표준 입력을 통해 한 줄을 입력받고 줄바꿈 등 양쪽 공백을 정리합니다.
        expr = sys.stdin.readline().strip()
        # 정답을 구해서 모니터에 출력해 줍니다.
        print(solution(expr))
    except Exception:
        pass  # 돌발 예외가 일어나더라도 조용히 종료시킵니다.

File: Week2/과제/완전_탐색/수식_최대화/Solution_Maximum.cpp
// 1. 필요한 C++ 표준 라이브러리 헤더 파일들을 불러옵니다.
#include <algorithm> // 정렬(sort) 및 순열(next_permutation) 기능을 사용하기 위한 라이브러리입니다.
#include <iostream>  // 표준 입출력(cin, cout)을 담당하는 라이브러리입니다.
#include <string>    // 문자열을 다루기 위한 라이브러리입니다.
#include <vector>    // 동적 배열(크기가 자유롭게 변하는 배열)인 vector를 사용하기 위한 라이브러리입니다.

// 매번 cout << endl 할 때 화면을 지우고 줄바꿈을 빠르게 처리하기 위한 매크로 정의입니다.
#define endl "\n"

using namespace std; // 표준 라이브러리의 이름공간(std)을 생략하고 편리하게 쓸 수 있도록 설정합니다.

/**
 * [두 숫자와 연산자를 받아서 계산 결과를 돌려주는 함수]
 * @param a  : 계산할 첫 번째 숫자 (큰 숫자도 처리할 수 있게 long long 자료형 사용)
 * @param b  : 계산할 두 번째 숫자 (long long 자료형)
 * @param op : 계산에 사용할 연산자 문자 ('+', '-', '*' 중 하나)
 * @return   : 계산된 결과 값
 */
long long calc(long long a, long long b, char op)
{
    // 연산자가 더하기('+')이면 두 수를 더해서 반환합니다.
    if (op == '+')
        return a + b;
    // 연산자가 빼기('-')이면 첫 번째 수에서 두 번째 수를 빼서 반환합니다.
    if (op == '-')
        return a - b;
    // 연산자가 곱하기('*')이면 두 수를 곱해서 반환합니다.
    if (op == '*')
        return a * b;
    // 그 외 잘못된 연산자가 오면 0을 반환합니다.
    return 0;
}

/**
 * [수식을 분석하고 연산자 우선순위를 조합하여 계산한 절댓값 중 최댓값을 찾는 함수]
 * @param expression : 계산할 원본 수식 문자열 (예: "100-200*300-500+20")
 * @return           : 연산자 우선순위를 바꿨을 때 나올 수 있는 가장 큰 절댓값
 */
long long solution(string expression)
{
    // 수식에서 분리해낸 숫자들을 순서대로 저장할 공간입니다.
    vector<long long> nums;
    // 수식에서 분리해낸 연산자들('+', '-', '*')을 순서대로 저장할 공간입니다.
    vector<char> ops;

    // 숫자를 자릿수별로 이어 붙여서 임시로 보관할 문자열 변수입니다.
    string temp = "";
    
    // 원본 수식 문자열을 한 글자씩 처음부터 끝까지 확인합니다.
    for (char c : expression)
    {
        // 현재 글자가 숫자(0~9)인 경우
        if (isdigit(c))
        {
            temp += c; // 임시 문자열에 숫자를 이어서 덧붙입니다. (예: '1', '0' -> "10")
        }
        // 현재 글자가 숫자가 아니라 연산자인 경우
        else
        {
            // 지금까지 이어 붙였던 임시 문자열을 숫자로 변환(stoll: String to Long Long)해서 nums 배열에 넣습니다.
            nums.push_back(stoll(temp));
            // 만난 연산자 문자를 ops 배열에 차례대로 넣습니다.
            ops.push_back(c);
            // 다음 숫자를 새로 만들기 위해 임시 문자열을 다시 빈 칸으로 비워둡니다.
            temp = "";
        }
    }
    // 수식이 끝났을 때 마지막에 남아있던 숫자를 정수로 변환하여 숫자가 저장된 배열에 마저 넣어줍니다.
    nums.push_back(stoll(temp));

    // 사용 가능한 연산자들의 종류를 배열에 담아둡니다.
    vector<char> priority = {'*', '+', '-'};
    // 모든 연산자의 조합(순열)을 빠짐없이 만들기 위해, 시작 전 연산자 배열을 오름차순 정렬해 둡니다.
    sort(priority.begin(), priority.end());

    // 연산 결과의 최대 절댓값을 기록해 둘 변수입니다. 0으로 시작합니다.
    long long max_val = 0;

    // next_permutation을 활용해 연산자들의 모든 우선순위 순서(예: * -> + -> - 등)를 돌아가며 적용합니다.
    do
    {
        // 원본 숫자 배열과 연산자 배열을 훼손하지 않기 위해 복사본을 만들어 연산에 사용합니다.
        vector<long long> temp_nums = nums;
        vector<char> temp_ops = ops;

        // 현재 조합으로 정해진 연산자 순서(우선순위가 가장 높은 연산자부터 순서대로)대로 계산을 시작합니다.
        for (char p_op : priority)
        {
            // 임시 연산자 배열을 처음부터 끝까지 순회합니다.
            for (int i = 0; i < temp_ops.size();)
            {
                // 현재 순서의 연산자(p_op)와 현재 위치에 있는 연산자가 일치하는 경우
                if (temp_ops[i] == p_op)
                {
                    // 해당 위치(i)의 숫자와 그 바로 다음(i+1) 위치의 숫자를 가져와서 연산자로 계산합니다.
                    long long res = calc(temp_nums[i], temp_nums[i + 1], p_op);
                    
                    // 계산된 결과 값을 첫 번째 숫자가 있던 자리에 덮어씌웁니다.
                    temp_nums[i] = res;
                    
                    // 사용한 두 번째 숫자(i+1번째)는 연산이 끝났으므로 배열에서 지워버립니다.
                    temp_nums.erase(temp_nums.begin() + i + 1);
                    
                    // 사용이 끝난 연산자(i번째)도 연산자 배열에서 삭제합니다.
                    temp_ops.erase(temp_ops.begin() + i);
                    
                    // 요소를 지웠으므로 배열의 크기가 줄어들었습니다. 
                    // 따라서 i를 증가시키지 않고 그 자리에서 다음 연산자를 계속 검사합니다.
                }
                // 일치하지 않는 경우
                else
                {
                    i++; // 다음 연산자를 확인하기 위해 인덱스를 1 증가시킵니다.
                }
            }
        }
        // 수식을 모두 계산하고 남은 최종 값(temp_nums[0])의 절댓값(abs)을 구한 뒤, 
        // 기존의 최댓값(max_val)과 비교하여 더 큰 값으로 갱신해 줍니다.
        max_val = max(max_val, abs(temp_nums[0]));
        
    } while (next_permutation(priority.begin(), priority.end())); // 다음 연산자 순서 조합이 있다면 계속 반복합니다.

    // 찾은 절댓값의 최대 결과를 최종 반환합니다.
    return max_val;
}

// 프로그램의 시작점이 되는 메인 함수입니다.
int main()
{
    string s; // 입력받을 수식 문자열을 선언합니다.
    cin >> s; // 사용자로부터 수식을 입력받습니다. (예: "100-200*300-500+20")
    
    // 계산 함수를 호출하고 그 결과를 화면에 출력한 뒤 줄을 바꿉니다.
    cout << solution(s) << endl;
    
    return 0; // 프로그램을 정상적으로 종료합니다.
}